Handle exception in `GET /project/{project_id}/dataset/{dataset_id}/status/`

NOTE: this applies to both V1 and V2 API.

This is from the fractal-server logs:
```python traceback
[2024-02-29 07:29:23 +0000] [5783] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 412, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/routing.py", line 758, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/routing.py", line 778, in app
    await route.handle(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/routing.py", line 299, in handle
    await self.app(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/routing.py", line 79, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/starlette/routing.py", line 74, in app
    response = await func(request)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/fastapi/routing.py", line 299, in app
    raise e
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/fastapi/routing.py", line 294, in app
    raw_response = await run_endpoint_function(
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/site-packages/fractal_server/app/routes/api/v1/dataset.py", line 511, in get_workflowtask_status
    history = json.load(f)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/somewhere/anaconda3/envs/fractal-server-1.4.9/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Apparently the history file may not include a valid JSON (why? TBD). Let's at least catch the exception properly within a try/except and raise a 422 error or similar.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,8 @@
 * Dependencies:
     * Make `clusterfutures` and `cloudpickle` required dependencies (\#1255).
     * Remove `slurm` extra from package (\#1255).
+* API:
+    * Handle invalid history file in `GET /project/{project_id}/dataset/{dataset_id}/status/` (\#1259).
 
 # 1.4.9
 

--- a/fractal_server/app/routes/api/v1/dataset.py
+++ b/fractal_server/app/routes/api/v1/dataset.py
@@ -1,4 +1,5 @@
 import json
+from json.decoder import JSONDecodeError
 from pathlib import Path
 from typing import Optional
 
@@ -511,6 +512,12 @@ async def get_workflowtask_status(
                 history = json.load(f)
         except FileNotFoundError:
             history = []
+        except JSONDecodeError:
+            raise HTTPException(
+                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                detail="History file does not include a valid JSON.",
+            )
+
         for history_item in history:
             wftask_id = history_item["workflowtask"]["id"]
             wftask_status = history_item["status"]

--- a/tests/04_api/test_history.py
+++ b/tests/04_api/test_history.py
@@ -324,3 +324,50 @@ async def test_assemble_history_failed_job_fail(
     assert "Cannot identify the failed task" in caplog.text
     debug(history)
     assert history == tmp_history
+
+
+async def test_json_decode_error(
+    db,
+    MockCurrentUser,
+    tmp_path,
+    project_factory,
+    task_factory,
+    dataset_factory,
+    workflow_factory,
+    job_factory,
+    client,
+):
+
+    history = "not a json"
+    working_dir = tmp_path / "working_dir"
+    working_dir.mkdir()
+    with (working_dir / HISTORY_FILENAME).open("w") as f:
+        f.write(history)
+
+    async with MockCurrentUser() as user:
+        project = await project_factory(user)
+        task = await task_factory(name="task1", source="task1")
+        workflow = await workflow_factory(project_id=project.id, name="WF")
+        await _workflow_insert_task(
+            workflow_id=workflow.id, task_id=task.id, db=db
+        )
+        input_dataset = await dataset_factory(project_id=project.id)
+        output_dataset = await dataset_factory(
+            project_id=project.id, history=history
+        )
+        await job_factory(
+            project_id=project.id,
+            workflow_id=workflow.id,
+            input_dataset_id=input_dataset.id,
+            output_dataset_id=output_dataset.id,
+            working_dir=str(working_dir),
+        )
+
+        # Test get_workflowtask_status endpoint
+        res = await client.get(
+            f"api/v1/project/{project.id}/dataset/{output_dataset.id}/status/"
+        )
+        assert res.status_code == 422
+        assert res.json()["detail"] == (
+            "History file does not include a valid JSON."
+        )

--- a/tests/04_api/test_history.py
+++ b/tests/04_api/test_history.py
@@ -338,7 +338,7 @@ async def test_json_decode_error(
     client,
 ):
 
-    history = "not a json"
+    history = "NOT A VALID JSON"
     working_dir = tmp_path / "working_dir"
     working_dir.mkdir()
     with (working_dir / HISTORY_FILENAME).open("w") as f:
@@ -346,7 +346,7 @@ async def test_json_decode_error(
 
     async with MockCurrentUser() as user:
         project = await project_factory(user)
-        task = await task_factory(name="task1", source="task1")
+        task = await task_factory(name="task", source="task")
         workflow = await workflow_factory(project_id=project.id, name="WF")
         await _workflow_insert_task(
             workflow_id=workflow.id, task_id=task.id, db=db
@@ -363,7 +363,6 @@ async def test_json_decode_error(
             working_dir=str(working_dir),
         )
 
-        # Test get_workflowtask_status endpoint
         res = await client.get(
             f"api/v1/project/{project.id}/dataset/{output_dataset.id}/status/"
         )

